lg69t/firmware_tool.py

```diff
@@ -240,21 +264,44 @@ def main():
     args = parser.parse_args()
 
     port_name = args.port
+    p1fw_path = args.p1fw
     gnss_bin_path = args.gnss
     app_bin_path = args.app
     should_send_reboot = not args.manual_reboot
 
-    if app_bin_path is None and gnss_bin_path is None:
-        print('You must specify gnss or app files to upgrade.')
+    if p1fw_path is None and app_bin_path is None and gnss_bin_path is None:
+        print('You must specify p1fw file, gnss file, or app file to upgrade.')
         sys.exit(1)
 
     print(f"Starting upgrade on device {port_name}.")
 
-    if gnss_bin_path is not None:
+    fw_files = {}
+    if p1fw_path is not None:
+        try:
+            with ZipFile(p1fw_path, 'r') as f:
```

Comment:
You shouldn't extract the zip data to a file, you can access it directly in memory without creating new files: https://stackoverflow.com/a/10909016

Reply:
The basic logic I'd do is that I'd change the `Upgrade` function to take a file handle instead of a file path.

Then I'd do something like   (pseudo code):

```python
app_bin_fd = None
if app_zip_file:
  app_bin_fd  = load_zip
  if app_bin_path:
    print("warning ignoring")
elif app_bin_path:
  app_bin_fd  = open(app_bin_path)
  
 if app_bin_fd:
   Upgrade(app_bin_fd, UpgradeType.APP)
 ```
 same thing for gnss firmware.

Reply:
Updated so that we're accessing the files from the compressed folder in memory. The one key difference from your listed approach is that I extract both of the files from the compressed folder into memory at the same time.